src/mlpack/methods/cf/ncf_impl.hpp

```diff
@@ -0,0 +1,371 @@
+/**
+ * @file ncf.hpp
+ * @author Haritha Nair
+ *
+ * Implementation of Neural Collaborative Filtering.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_NCF__NCF_IMPL_HPP
+#define MLPACK_METHODS_NCF_IMPL_NCF_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "ncf.hpp"
+
+namespace mlpack {
+namespace cf {
+/**
+ * Construct the NCF object using the desired algorithm and optimizer.
+ */
+template<typename OptimizerType>
+NCF::NCF(arma::mat& dataset,
+         std::string algorithm,
+         OptimizerType& optimizer,
+         const size_t embedSize,
+         const size_t neg,
+         const size_t epochs,
+         bool implicit):
+    dataset(dataset),
+    network(ann::FFN<ann::NegativeLogLikelihood<>,
```

Comment:
There was some error regarding having to initialize network in constructor. But now since CreateModel() methods are in constructor I think I can remove it.